File: packages/python_backend/app/utils/storage/project_storage.py

import json
import os
import time
from pathlib import Path
from typing import Any, Dict, Type, TypeVar, Optional, Union, List
from pydantic import BaseModel, ValidationError, RootModel
from datetime import datetime, timezone

# Import the authoritative schemas
from app.schemas.project_schemas import Project, ProjectFile
from app.schemas.ai_file_change_schemas import AIFileChangeRecord

WORKSPACE_ROOT = Path.cwd()
DATA_DIR = WORKSPACE_ROOT / "data" / "python_project_storage"

ProjectsStorageModel = RootModel[Dict[int, Project]]
ProjectFilesStorageModel = RootModel[Dict[int, ProjectFile]]
AIFileChangesStorageModel = RootModel[Dict[int, AIFileChangeRecord]]

T = TypeVar("T", bound=BaseModel)
DictModel = TypeVar("DictModel", bound=RootModel)

def get_projects_index_path() -> Path: return DATA_DIR / "projects.json"
def get_project_data_dir(project_id: int) -> Path: return DATA_DIR / "project_data" / str(project_id)
def get_project_files_path(project_id: int) -> Path: return get_project_data_dir(project_id) / "files.json"
def get_ai_file_changes_path(project_id: int) -> Path: return get_project_data_dir(project_id) / "ai_file_changes.json"

async def _ensure_dir_exists(dir_path: Path) -> None:
    try: dir_path.mkdir(parents=True, exist_ok=True)
    except Exception as e: raise IOError(f"Failed to ensure directory exists: {dir_path}") from e

async def _read_validated_json(
    file_path: Path,
    model_type: Union[Type[T], Type[DictModel]],
    default_value: Union[Dict, List]
) -> Union[T, Dict[int, Any], List[Any]]:
    try:
        await _ensure_dir_exists(file_path.parent)
        if not file_path.exists(): return default_value

        with open(file_path, "r", encoding="utf-8") as f: file_content = f.read()
        if not file_content.strip(): return default_value

        json_data = json.loads(file_content)

        if hasattr(model_type, 'model_fields') and 'root' in model_type.model_fields and \
           isinstance(model_type.model_fields['root'].annotation, type(RootModel[Any])) and \
           hasattr(model_type.model_fields['root'].annotation, '__args__') and \
           len(model_type.model_fields['root'].annotation.__args__) > 0 and \
           isinstance(model_type.model_fields['root'].annotation.__args__[0], type(Dict[int, Any])):

            value_model_type_hint = model_type.model_fields['root'].annotation.__args__[0]
            if hasattr(value_model_type_hint, '__args__') and len(value_model_type_hint.__args__) == 2:
                value_model_type = value_model_type_hint.__args__[1]
                parsed_dict = {}
                if isinstance(json_data, dict):
                    for str_k, v_data in json_data.items():
                        try:
                            int_k = int(str_k)
                            parsed_dict[int_k] = value_model_type.model_validate(v_data)
                        except (ValueError, ValidationError):
                            continue 
                return model_type(root=parsed_dict).root
            else:
                parsed_model = model_type.model_validate(json_data)
                return parsed_model.root
        elif issubclass(model_type, RootModel):
            parsed_model = model_type.model_validate(json_data)
            return parsed_model.root
        elif issubclass(model_type, BaseModel):
            return model_type.model_validate(json_data)
        else: raise TypeError("model_type must be a Pydantic BaseModel or RootModel subclass")

    except FileNotFoundError: return default_value
    except json.JSONDecodeError: return default_value
    except ValidationError: return default_value
    except Exception as e: raise IOError(f"Failed to read/parse JSON: {file_path}") from e

async def _write_validated_json(
    file_path: Path,
    data: Union[BaseModel, RootModel, Dict[int, Any], List],
    model_type: Union[Type[T], Type[DictModel]]
) -> Union[T, Dict[int, Any], List[Any]]:
    try:
        validated_data_to_serialize: Any
        output_data_structure: Any

        if isinstance(data, BaseModel):
            validated_data_to_serialize = data.model_dump(mode='json')
            output_data_structure = data
        elif isinstance(data, RootModel):
            internal_data = data.root
            if isinstance(internal_data, dict):
                validated_data_to_serialize = {str(k): v.model_dump(mode='json') if isinstance(v, BaseModel) else v for k,v in internal_data.items()}
            else:
                validated_data_to_serialize = data.model_dump(mode='json')
            output_data_structure = data
        elif isinstance(data, dict):
            temp_root_model = model_type(root={k:v for k,v in data.items()})
            internal_dict = temp_root_model.root
            validated_data_to_serialize = {str(k): v.model_dump(mode='json') if isinstance(v, BaseModel) else v for k,v in internal_dict.items()}
            output_data_structure = temp_root_model
        elif isinstance(data, list):
            parsed_model = model_type.model_validate(data)
            validated_data_to_serialize = parsed_model.model_dump(mode='json')
            output_data_structure = parsed_model
        else: raise TypeError("Data to write must be a Pydantic model, RootModel, dict, or list")

        await _ensure_dir_exists(file_path.parent)
        json_string = json.dumps(validated_data_to_serialize, indent=2)
        with open(file_path, "w", encoding="utf-8") as f: f.write(json_string)

        if isinstance(output_data_structure, RootModel): return output_data_structure.root
        return output_data_structure

    except ValidationError as e: raise e
    except Exception as e: raise IOError(f"Failed to write JSON: {file_path}") from e

class ProjectStorage:
    def __init__(self):
        self._last_timestamp: int = 0
        self._sequence_counter: int = 0
    
    async def read_projects(self) -> Dict[int, Project]:
        return await _read_validated_json(get_projects_index_path(), ProjectsStorageModel, {})

    async def write_projects(self, projects: Dict[int, Union[Project, Dict]]) -> Dict[int, Project]:
        return await _write_validated_json(get_projects_index_path(), projects, ProjectsStorageModel)

    async def read_project_files(self, project_id: int) -> Dict[int, ProjectFile]:
        return await _read_validated_json(get_project_files_path(project_id), ProjectFilesStorageModel, {})

    async def write_project_files(self, project_id: int, files: Dict[int, Union[ProjectFile, Dict]]) -> Dict[int, ProjectFile]:
        return await _write_validated_json(get_project_files_path(project_id), files, ProjectFilesStorageModel)

    async def read_ai_file_changes(self, project_id: int) -> Dict[int, AIFileChangeRecord]:
        return await _read_validated_json(get_ai_file_changes_path(project_id), AIFileChangesStorageModel, {})

    async def write_ai_file_changes(self, project_id: int, changes: Dict[int, Union[AIFileChangeRecord, Dict]]) -> Dict[int, AIFileChangeRecord]:
        return await _write_validated_json(get_ai_file_changes_path(project_id), changes, AIFileChangesStorageModel)

    async def read_project_file(self, project_id: int, file_id: int) -> Optional[ProjectFile]:
        files = await self.read_project_files(project_id)
        return files.get(file_id)

    async def update_project_file(
        self, project_id: int, file_id: int, file_update_data: Dict[str, Any]
    ) -> ProjectFile:
        current_file = await self.read_project_file(project_id, file_id)
        if not current_file: raise FileNotFoundError(f"File not found: {file_id} in project {project_id}")

        if "updated" not in file_update_data or not isinstance(file_update_data["updated"], int):
                 file_update_data["updated"] = int(datetime.now(timezone.utc).timestamp() * 1000)

        updated_file_model = current_file.model_copy(update=file_update_data)

        try: validated_file_data = ProjectFile(**updated_file_model.model_dump())
        except ValidationError as e: raise e

        current_project_files = await self.read_project_files(project_id)
        current_project_files[file_id] = validated_file_data
        await self.write_project_files(project_id, current_project_files)
        return validated_file_data

    async def get_ai_file_change_by_id(self, project_id: int, change_id: int) -> Optional[AIFileChangeRecord]:
        changes = await self.read_ai_file_changes(project_id)
        return changes.get(change_id)

    async def save_ai_file_change(self, project_id: int, record: AIFileChangeRecord) -> None:
        changes = await self.read_ai_file_changes(project_id)
        changes[record.id] = record
        await self.write_ai_file_changes(project_id, changes)

    async def delete_project_data(self, project_id: int) -> None:
        dir_path = get_project_data_dir(project_id)
        try:
            if dir_path.exists(): import shutil; shutil.rmtree(dir_path)
        except FileNotFoundError: pass
        except Exception as e: raise IOError(f"Failed to delete dir: {dir_path}") from e

    def generate_id(self) -> int:
        current_timestamp = int(time.time() * 1000)
        if current_timestamp == self._last_timestamp:
            self._sequence_counter += 1
            unique_id = current_timestamp + self._sequence_counter
        else:
            self._last_timestamp = current_timestamp
            self._sequence_counter = 0
            unique_id = current_timestamp
        return unique_id

project_storage = ProjectStorage()

File: packages/server/src/utils/storage/project-storage.ts
// src/utils/project-storage.ts
import { z, ZodError, type ZodTypeAny } from 'zod'
import path from 'node:path'
import fs from 'node:fs/promises' // Using Node's fs promises
import { ProjectSchema, ProjectFileSchema, type ProjectFile } from 'shared/src/schemas/project.schemas'
import { AIFileChangesStorageSchema, type AIFileChangesStorage, type AIFileChangeRecord, AIFileChangeRecordSchema } from 'shared/src/schemas/ai-file-change.schemas'
import { normalizeToUnixMs } from '../parse-timestamp'

// Define the base directory for storing project data
// Adjust this path as needed, e.g., use an environment variable
const DATA_DIR = path.resolve(process.cwd(), 'data', 'project_storage')

// --- Schemas for Storage ---
// Store projects as a map (Record) keyed by projectId
export const ProjectsStorageSchema = z.record(z.number(), ProjectSchema)
export type ProjectsStorage = z.infer<typeof ProjectsStorageSchema>

// even though the keys are numbers, they are saved as string because that is default javascript behavior
export const ProjectFilesStorageSchema = z.record(z.string(), ProjectFileSchema)
export type ProjectFilesStorage = z.infer<typeof ProjectFilesStorageSchema>

// --- Path Helpers ---

/** Gets the absolute path to the main projects index file. */
function getProjectsIndexPath(): string {
  return path.join(DATA_DIR, 'projects.json')
}

/** Gets the absolute path to a specific project's directory. */
function getProjectDataDir(projectId: number): string {
  return path.join(DATA_DIR, 'project_data', projectId.toString())
}

/** Gets the absolute path to a specific project's files index file. */
function getProjectFilesPath(projectId: number): string {
  return path.join(getProjectDataDir(projectId), 'files.json')
}

/** Gets the absolute path to a specific project's AI file changes index file. */
function getProjectAIFileChangesPath(projectId: number): string {
  return path.join(getProjectDataDir(projectId), 'ai-file-changes.json')
}

// --- Core Read/Write Functions ---

/** Ensures the base data directory and project-specific directories exist. */
async function ensureDirExists(dirPath: string): Promise<void> {
  try {
    await fs.mkdir(dirPath, { recursive: true })
  } catch (error: any) {
    // Ignore EEXIST error (directory already exists), re-throw others
    if (error.code !== 'EEXIST') {
      console.error(`Error creating directory ${dirPath}:`, error)
      throw new Error(`Failed to ensure directory exists: ${dirPath}`)
    }
  }
}

/**
 * Reads and validates JSON data from a file.
 * @param filePath The absolute path to the JSON file.
 * @param schema The Zod schema to validate against.
 * @param defaultValue The value to return if the file doesn't exist.
 * @returns Validated data or the default value.
 */
async function readValidatedJson<T extends ZodTypeAny>(
  filePath: string,
  schema: T,
  defaultValue: z.infer<T>
): Promise<z.infer<T>> {
  try {
    await ensureDirExists(path.dirname(filePath)) // Ensure parent dir exists
    const fileContent = await fs.readFile(filePath, 'utf-8')

    // Handle cases where fileContent might be empty or only whitespace
    if (fileContent.trim() === '') {
      console.warn(`File is empty or contains only whitespace: ${filePath}. Returning default value.`);
      return defaultValue;
    }

    let jsonData = JSON.parse(fileContent) // This can throw SyntaxError

    console.log('jsonData', jsonData)

    // If the schema is a ZodRecord with numeric keys, transform string keys to numbers
    if (schema instanceof z.ZodRecord && schema.keySchema instanceof z.ZodNumber) {
      const transformedData: Record<number, unknown> = {};
      for (const key in jsonData) {
        if (Object.prototype.hasOwnProperty.call(jsonData, key)) {
          const numKey = Number(key);
          if (!isNaN(numKey)) {
            transformedData[numKey] = jsonData[key];
          } else {
            // If the key is not a valid number for a schema expecting numeric keys,
            // log a warning and omit this key-value pair from the transformed data.
            // This ensures that 'transformedData' only contains numeric keys,
            // allowing Zod validation for z.record(z.number(), ...) to pass if values are correct.
            console.warn(`Omitting non-numeric key "${key}" from object in ${filePath} as schema expects numeric keys.`);
          }
        }
      }
      jsonData = transformedData;
    }

    const validationResult = await schema.safeParseAsync(jsonData)

    if (!validationResult.success) {
      console.error(`Zod validation failed reading ${filePath}:`, validationResult.error.errors)
      console.warn(`Returning default value due to validation failure for ${filePath}.`)
      return defaultValue
    }
    return validationResult.data
  } catch (error: any) {
    if (error.code === 'ENOENT') {
      return defaultValue
    }
    // Specifically catch JSON parsing errors
    if (error instanceof SyntaxError) {
      console.error(`JSON Parse error in ${filePath}:`, error.message);
      console.warn(`Returning default value due to JSON parsing error for ${filePath}.`);
      return defaultValue;
    }
    console.error(`Error reading or parsing JSON from ${filePath}:`, error)
    throw new Error(`Failed to read/parse JSON file at ${filePath}. Reason: ${error.message}`)
  }
}

/**
 * Validates data and writes it to a JSON file.
 * @param filePath The absolute path to the JSON file.
 * @param data The data to write.
 * @param schema The Zod schema to validate against.
 * @returns The validated data that was written.
 */
async function writeValidatedJson<T extends ZodTypeAny>(
  filePath: string,
  data: unknown, // Accept unknown initially for validation
  schema: T
): Promise<z.infer<T>> {
  try {
    // 1. Validate data first
    const validationResult = await schema.safeParseAsync(data)
    if (!validationResult.success) {
      console.error(`Zod validation failed before writing to ${filePath}:`, validationResult.error.errors)
      throw new ZodError(validationResult.error.errors)
    }
    const validatedData = validationResult.data

    // 2. Ensure directory exists
    await ensureDirExists(path.dirname(filePath))

    // 3. Stringify and write
    const jsonString = JSON.stringify(validatedData, null, 2) // Pretty print
    await fs.writeFile(filePath, jsonString, 'utf-8')

    // console.log(`Successfully validated and wrote JSON to: ${filePath}`);
    return validatedData
  } catch (error: any) {
    console.error(`Error writing JSON to ${filePath}:`, error)
    if (error instanceof ZodError) {
      throw error // Re-throw Zod errors
    }
    throw new Error(`Failed to write JSON file at ${filePath}. Reason: ${error.message}`)
  }
}

// --- Specific Data Accessors ---

export const projectStorage = {
  /** Reads the main projects data file. */
  async readProjects(): Promise<ProjectsStorage> {
    return readValidatedJson(getProjectsIndexPath(), ProjectsStorageSchema, {})
  },

  /** Writes the main projects data file. */
  async writeProjects(projects: ProjectsStorage): Promise<ProjectsStorage> {
    return writeValidatedJson(getProjectsIndexPath(), projects, ProjectsStorageSchema)
  },

  /** Reads a specific project's file data. */
  async readProjectFiles(projectId: number): Promise<ProjectFilesStorage> {
    return readValidatedJson(getProjectFilesPath(projectId), ProjectFilesStorageSchema, {})
  },

  /** Writes a specific project's file data. */
  async writeProjectFiles(projectId: number, files: ProjectFilesStorage): Promise<ProjectFilesStorage> {
    return writeValidatedJson(getProjectFilesPath(projectId), files, ProjectFilesStorageSchema)
  },

  /** Reads a specific project's AI file changes data. */
  async readAIFileChanges(projectId: number): Promise<AIFileChangesStorage> {
    return readValidatedJson(getProjectAIFileChangesPath(projectId), AIFileChangesStorageSchema, {})
  },

  /** Writes a specific project's AI file changes data. */
  async writeAIFileChanges(projectId: number, changes: AIFileChangesStorage): Promise<AIFileChangesStorage> {
    return writeValidatedJson(getProjectAIFileChangesPath(projectId), changes, AIFileChangesStorageSchema)
  },

  /**
   * Adds or updates a specific AI file change record within a project.
   * @param projectId The ID of the project.
   * @param changeData The AI file change data to add or update.
   * @returns The validated and saved AIFileChangeRecord data.
   */
  async saveAIFileChange(projectId: number, changeData: AIFileChangeRecord): Promise<AIFileChangeRecord> {
    const validationResult = await AIFileChangeRecordSchema.safeParseAsync(changeData);
    if (!validationResult.success) {
      console.error(`Zod validation failed for AI file change in project ${projectId}:`, validationResult.error.errors);
      throw new ZodError(validationResult.error.errors);
    }
    const validatedChangeData = validationResult.data;

    const currentChanges = await this.readAIFileChanges(projectId);
    currentChanges[validatedChangeData.id] = validatedChangeData;

    await this.writeAIFileChanges(projectId, currentChanges);
    return validatedChangeData;
  },

  /**
   * Retrieves a specific AI file change record by its ID from a project.
   * @param projectId The ID of the project.
   * @param aiFileChangeId The ID of the AI file change record.
   * @returns The AIFileChangeRecord or undefined if not found.
   */
  async getAIFileChangeById(projectId: number, aiFileChangeId: number): Promise<AIFileChangeRecord | undefined> {
    const changes = await this.readAIFileChanges(projectId);
    return changes[aiFileChangeId];
  },

  /**
   * Updates a specific file within a project.
   * Reads the project's files, merges the new data, sets the `updated` timestamp,
   * validates the complete object, and writes the entire file map back.
   * @param projectId The ID of the project containing the file.
   * @param fileId The ID of the file to update.
   * @param fileData A partial object of file properties to update. Fields like id, projectId, created, and updated are ignored.
   * @returns The validated and saved ProjectFile data.
   * @throws ZodError if validation fails.
   * @throws Error if the file is not found or if reading/writing fails.
   */
  async updateProjectFile(
    projectId: number,
    fileId: number,
    fileData: Partial<Omit<ProjectFile, 'id' | 'projectId' | 'created' | 'updated'>>
  ): Promise<ProjectFile> {
    const currentFiles = await this.readProjectFiles(projectId)
    const currentFile = currentFiles[fileId]
    if (!currentFile) {
      throw new Error(`File not found: ${fileId} in project ${projectId}`)
    }

    const updatedFileObject = {
      ...currentFile,
      ...fileData,
      updated: Date.now() // Always set a new update timestamp in milliseconds
    }

    // ProjectFileSchema is expected to be updated to match Python's schema
    // with `created` and `updated` as number fields.
    const validationResult = await ProjectFileSchema.safeParseAsync(updatedFileObject)

    if (!validationResult.success) {
      console.error(`Zod validation failed for file ${fileId} in project ${projectId}:`, validationResult.error.errors)
      throw new ZodError(validationResult.error.errors)
    }
    const validatedFileData = validationResult.data

    currentFiles[fileId] = validatedFileData
    await this.writeProjectFiles(projectId, currentFiles)
    return validatedFileData
  },

  async readProjectFile(projectId: number, fileId: number): Promise<ProjectFile | undefined> {
    const files = await this.readProjectFiles(projectId)
    return files[fileId]
  },

  /** Deletes a project's data directory. */
  async deleteProjectData(projectId: number): Promise<void> {
    const dirPath = getProjectDataDir(projectId)
    try {
      // Check if directory exists before attempting removal
      await fs.access(dirPath) // Throws if doesn't exist
      await fs.rm(dirPath, { recursive: true, force: true }) // Remove dir and contents
    } catch (error: any) {
      if (error.code === 'ENOENT') {
        console.warn(`Project data directory not found, nothing to delete: ${dirPath}`)
        // Not an error condition if we're just ensuring it's gone
      } else {
        console.error(`Error deleting project data directory ${dirPath}:`, error)
        throw new Error(`Failed to delete project data directory: ${dirPath}. Reason: ${error.message}`)
      }
    }
  },

  /** Generates a simple unique ID (replace with more robust method if needed) */
  generateId: (): number => {
    return normalizeToUnixMs(new Date())
  }
}